fix: scheduled upgrade details state

pkoutsovasilis opened the pull request:
<!-- Type of change
Please label this PR with one of the following labels, depending on the scope of your change:
- Bug
- Enhancement
- Breaking change
- Deprecation
- Cleanup
- Docs
-->

## What does this PR do?

<!-- Mandatory
Explain here the changes you made on the PR. Please explain the WHAT: patterns used, algorithms implemented, design architecture, message processing, etc.
-->

This PR fixes how Elastic Agent persists and reports upgrade details across check-ins, restarts, cancellations, and expired scheduled upgrade actions.

Specifically:
- (https://github.com/elastic/elastic-agent/commit/95ab8574440caafe06d2a1d25a1660fe21db2c35) Ensures the `Coordinator` is initialized with the correct upgrade details, either from the upgrade marker file or queued Fleet actions. 
- (https://github.com/elastic/elastic-agent/commit/95ab8574440caafe06d2a1d25a1660fe21db2c35, https://github.com/elastic/elastic-agent/commit/cabe8c7690a20bf0b6059f3b21df92a5043ba291) Extends the `ActionDispatcher` logic to:
  - Deduplicate upgrade actions from the received fleetgateway actions by keeping only the first encountered upgrade in input order.
  - Correctly track when upgrade details need to be updated (new upgrade actions, cancellations, expirations, etc.).
  - Correctly handle upgrade action cancellations.
  - Correctly handle expired scheduled upgrade actions and keep persisting them until explicitly canceled or replaced by a newer upgrade action.
  - Correctly handle retried upgrade actions.
  - Correctly update upgrade details when an upgrade action is scheduled for a retry.
  - Correctly update the upgrade details to `nil` when an upgrade action is dispatched without an error.
- (https://github.com/elastic/elastic-agent/commit/a16a9f387ea00a98bd83c2506ff3504428cf616b) Refactors and extends related dispatcher tests to include an actual action queue and cover new flows (queueing, canceling, expiring).
- (https://github.com/elastic/elastic-agent/commit/3e7324e974609147223ff73ed3eaa0b905c03695) Adds new **integration tests** under `testing/integration/ess/scheduled_upgrade_details_test.go` to verify that scheduled upgrade details are:
  - Reported correctly when received.
  - Reported correctly when expired.
  - Preserved across Agent restarts.
  - Cleared on cancel.

All the new business logic is captured by https://github.com/elastic/elastic-agent/commit/95ab8574440caafe06d2a1d25a1660fe21db2c35 (+234 -126 lines changed) and https://github.com/elastic/elastic-agent/commit/cabe8c7690a20bf0b6059f3b21df92a5043ba291 (+174 -20 lines changed)

## Why is it important?

<!-- Mandatory
Explain here the WHY, or the rationale/motivation for the changes.
-->

Previously, scheduled upgrade details could be lost or incorrectly reported:
- They were not consistently preserved across Agent restarts.
- Cancel actions did not clear the reported upgrade details.
- Expired scheduled upgrade actions were treated as ephemeral and not persisted.

These inconsistencies caused confusion in Fleet UI and sometimes left users unable to perform upgrades.  
With this fix, users now get a consistent and accurate view of upcoming, active, or failed upgrades across the entire Agent lifecycle.

## Checklist

<!-- Mandatory
Add a checklist of things that are required to be reviewed in order to have the PR approved

List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply. (Just in case, strikethrough uses two tildes. ~~Scratch this.~~)
-->

- [x] I have read and understood the [pull request guidelines](https://github.com/elastic/elastic-agent/blob/main/CONTRIBUTING.md#pull-request-guidelines) of this project.
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)
- [x] I have added an integration test or an E2E test

## Disruptive User Impact

<!--
Will the changes introduced by this PR cause disruption to users in any way? If so, please describe what changes users
could make on their end to nullify or minimize this disruption. Consider impacts in related systems, not just directly
when using Elastic Agent.
-->

None expected. This change only improves correctness of reported upgrade details.  
Users upgrading to this version will see more consistent and reliable upgrade state reporting in Fleet.


## How to test this PR locally

<!-- Recommended
Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->
1. Run the new integration tests
2. Or build the code of this PR
  * Provision a 9.1.2 cloud stack
  * `AGENT_PACKAGE_VERSION=9.1.1 EXTERNAL=true SNAPSHOT=true PLATFORMS="linux/arm64" PACKAGES="tar.gz" mage package` (ty @ycombinator for the proposal)
  * Install and enroll the Agent to Fleet
  * Schedule an upgrade → check it is reported.
  * Restart the Agent → check details persist.
  * Cancel the upgrade → check details clear.

https://github.com/user-attachments/assets/92545d1e-da61-4131-8760-04311657f47f

Also I tested the approach mentioned [here](https://github.com/elastic/elastic-agent/issues/8778#issuecomment-3133874202) to force update an agent and everything is reflected correctly to the upgrade details

https://github.com/user-attachments/assets/d7f85c50-e451-405a-a14e-fc9feaa3375f

(PS: For the needs of the videos above I compiled this PR with pseudo prior version 9.1.1)

## Related issues

<!-- Recommended
Link related issues below. Insert the issue link or reference after the word "Closes" if merging this should automatically close it.

- Closes #123
- Relates #123
- Requires #123
- Superseds #123
-->
- Closes https://github.com/elastic/elastic-agent/issues/8778
- Closes https://github.com/elastic/elastic-agent/issues/9434

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/elastic-agent/builds/25711)
* Commit: 0a028ada6ff55723af6ac915104321e2a82ad808

### Failed CI Steps
* [check-ci](https://buildkite.com/elastic/elastic-agent/builds/25711#0198e18a-ae16-4404-b44b-fdb8bfd7dbd3)

cc @pkoutsovasilis

<!--buildkite-pr-comment-elastic-agent
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-agent/builds/25711","number":25711,"commit":"0a028ada6ff55723af6ac915104321e2a82ad808"}],"number":25711}
buildkite-pr-comment-->

blakerouse commented on internal/pkg/agent/application/dispatcher/dispatcher.go:
```diff
@@ -192,70 +203,165 @@ func detectTypes(actions []fleetapi.Action) []string {
+	ad.log.Debugf("Expired upgrade actions (won't be removed from the queue): %v", expiredUpgradeActions)
+
+	for _, expiredUpgradeAction := range expiredUpgradeActions {
+		startTime, err := expiredUpgradeAction.StartTime()
```
Can you add a comment here saying that at this point in the code there really should be a `StartTime` on the action, because these are expired upgrade actions. I just want to make warnings in the code clear, because before this change it was not clear.

pkoutsovasilis replied:
have a look here 6f75b29b772af12fe38091555902a056c14b9f35

blakerouse commented on internal/pkg/agent/application/dispatcher/dispatcher.go:
```diff
@@ -192,70 +203,165 @@ func detectTypes(actions []fleetapi.Action) []string {
+	upgradeDetails := GetScheduledUpgradeDetails(ad.log, ad.queue.Actions(), ts)
+
+	if ad.lastUpgradeDetailsIsSet && ad.lastUpgradeDetails.Equals(upgradeDetails) {
+		// calculate
```
```suggestion
		// already the same details; do nothing
```

pkoutsovasilis replied:
changed in 6f75b29b772af12fe38091555902a056c14b9f35

blakerouse reviewed (commented):
The code looks good, just a two nits that I found.

Overall the testing looks good. The integration testing is good. I do have a request for an extra integration test. I didn't see a test case for scheduling an upgrade action, the action running, it failing, then scheduling a new action not scheduled, that replaces that action with the same version and URI. I believe that could be added with an upgrade image behind a transparent proxy, where the first action always fails to download, and then the following up action it allows it to succeed.

ycombinator commented on internal/pkg/agent/application/dispatcher/dispatcher.go:
```diff
@@ -290,86 +396,64 @@ func (ad *ActionDispatcher) scheduleRetry(ctx context.Context, action fleetapi.R
+		if err != nil {
+			if !errors.Is(err, fleetapi.ErrNoExpiration) {
+				// this is not a non-expiring upgrade action
+				log.Errorf("failed to get expiration time for scheduled upgrade action [id = %s]: %v", upgradeAction.ID(), err)
```
Considering not all scheduled upgrade actions are expected to have an expiration, should we make this a warning instead of an error?

ycombinator replied:
Never mind, I got confused by the error message.  I see you are logging an error that occurred when getting the expiration time, but only when the action is _expected_ not to have an expiration.  The error is not that the action has no expiration time.

ycombinator commented on internal/pkg/agent/application/dispatcher/dispatcher.go:
```diff
@@ -192,70 +203,165 @@ func detectTypes(actions []fleetapi.Action) []string {
+	ad.log.Debugf("Expired non-upgrade actions: %v", expired)
+	ad.log.Debugf("Expired upgrade actions (won't be removed from the queue): %v", expiredUpgradeActions)
+
+	for _, expiredUpgradeAction := range expiredUpgradeActions {
```
Could you add a comment here explaining why we add back the expired upgrade actions to the queue?

pkoutsovasilis replied:
take a look here 6f75b29b772af12fe38091555902a056c14b9f35

ycombinator replied:
Thanks.  

Now I'm wondering why we only remove expired actions from the queue when we get a cancel or newer upgrade action removes them.  I understand we have to keep them in the queue to survive restarts so we can report that they expired in upgrade details but why not remove them immediately after that reporting? Why wait till we get a cnacel or newer upgrade action?

Anyway, I'm not going to block the improvements made by this PR on this question; it's more for my understanding.  Depending on the answer, I think it would be helpful to document it in the code somewhere but, again, that's for a follow up PR.

pkoutsovasilis replied:
no let's discuss this so all of us are in the same page and then we merge the fixes. So the reason these are added back in the queue and not removed, is to survive elastic-agent restarts; the next time agent start it will load the details from the marker or from the action queue. If expired actions are not there it will emit `nil` upgrade details so the expiration will be lost from the fleet UI. Does that explanation help?

pkoutsovasilis replied:
because if we did that and restarted this would result in our next checkin sending nil upgrade details and thus the Fleet UI wouldn't show the expiration anymore? happy to have a sync about

ycombinator replied:
Ah okay, I got it now.  We discussed this scenario in one of the team syncs.  We need the Fleet UI to keep showing the failed state from the upgrade details until the user takes some action.  Okay, makes sense now but maybe add a comment about this in the code so I don't ask this question again in 3 weeks 😅?

pkoutsovasilis replied:
I’ve added already a comment clarifying that these need to be persisted, which should be enough to prevent someone from altering this without checking in with the team first. For now, I’ll go ahead and merge this since it fixes the issue we have at hand. And if in 3 weeks you forget about it 😅 I’ll open a follow-up PR to extend the comment further and explicitly mention that this was a team decision to always show expired ones in Fleet UI until the user takes another upgrade action.

ycombinator reviewed (approved):
Code LGTM.  Just left a few minor suggestions.

pkoutsovasilis commented on internal/pkg/agent/application/dispatcher/dispatcher.go:
```diff
@@ -290,86 +396,64 @@ func (ad *ActionDispatcher) scheduleRetry(ctx context.Context, action fleetapi.R
-		}
-	}
-}
+// GetScheduledUpgradeDetails returns the upgrade details of the next upgrade action, if any. It also adjusts accordingly
```
have a look here 6f75b29b772af12fe38091555902a056c14b9f35

pkoutsovasilis commented on internal/pkg/agent/application/dispatcher/dispatcher.go:
```diff
@@ -192,70 +203,165 @@ func detectTypes(actions []fleetapi.Action) []string {
+			*upgradeDetailsNeedUpdate = true
+		}
+
+		continue
```
removed here c976e0699bc5c96ca110c815bf72d43627fda5ff

pkoutsovasilis commented:
@blakerouse @ycombinator while working on the unit test we agreed on, I ran into a few more issues (story of my life :bowtie:). That’s why I added a new commit (https://github.com/elastic/elastic-agent/commit/cabe8c7690a20bf0b6059f3b21df92a5043ba291) with additional business logic changes. I’ve updated the PR description, but to call them out specifically:

> Deduplicate upgrade actions from the received fleetgateway actions by keeping only the first encountered upgrade in input order.

We now keep only the first encountered upgrade action in input order. I’m not sure how/why fleetgateway can send multiple upgrade actions, but it causes problems. Example: if we get both an immediate upgrade and a scheduled upgrade, which one should set the upgrade details? And what if the scheduled one gets dispatched in the next window while the immediate one is still running? That feels like something that shouldn’t be allowed, but I’d like your thoughts @blakerouse @ycombinator 

> Correctly handle retried upgrade actions.

Previously ([code link](https://github.com/elastic/elastic-agent/blob/main/internal/pkg/agent/application/dispatcher/dispatcher.go#L236)), retried immediate upgrade actions (no expiration) were incorrectly treated as expired because `exp` was `0,` and `ts.After(exp)` was always true.

> Correctly update upgrade details when an upgrade action is scheduled for a retry.

The dispatcher now updates upgrade details on retry, including the error from the prior attempt. Unfortunately, Fleet UI still shows only the `UPG_SCHEDULED` tag without an error indicator (e.g. a warning/triangle), but the agent state correctly captures this - so I think this is fine.

> Correctly update the upgrade details to `nil` when an upgrade action is dispatched without an error.

Learned here that the coordinator upgrader can skip setting upgrade details (e.g. if the action is ignored because it’s the same version/URI). Without resetting, the last dispatcher-set upgrade details would “stick” to the agent and, since the action gets ack’ed, there’d be no cancellation path. Fix: whenever the dispatcher sees a dispatchable upgrade action, it sets upgrade details to `nil`. That way, if the upgrader doesn’t set new ones, we don’t end up with stale data.

---

Now, all that said, I think there’s another upcoming issue: in [PR #8767](https://github.com/elastic/elastic-agent/pull/8767), available rollback versions are becoming part of the upgrade details. That can create issues since the dispatcher knows nothing about which versions are available for rollback, so it could incorrectly set upgrade details with these missing. From what I understand now about how upgrade details flow through the code, rollback versions don’t really belong in upgrade details - they should probably be part of the check-in request body instead. Fleet UI could then combine those with upgrade details to decide whether to show a rollback option.

@pchila @ebeahan let's discuss about that - maybe I am missing something

blakerouse commented:
> @blakerouse @ycombinator while working on the unit test we agreed on, I ran into a few more issues (story of my life :bowtie:). That’s why I added a new commit ([cabe8c7](https://github.com/elastic/elastic-agent/commit/cabe8c7690a20bf0b6059f3b21df92a5043ba291)) with additional business logic changes. I’ve updated the PR description, but to call them out specifically:

@pkoutsovasilis Great that these extra issues where caught now, with added unit tests. The changes adding in that commit look good, very well commented.

> 
> > Deduplicate upgrade actions from the received fleetgateway actions by keeping only the first encountered upgrade in input order.
> 
> We now keep only the first encountered upgrade action in input order. I’m not sure how/why fleetgateway can send multiple upgrade actions, but it causes problems. Example: if we get both an immediate upgrade and a scheduled upgrade, which one should set the upgrade details? And what if the scheduled one gets dispatched in the next window while the immediate one is still running? That feels like something that shouldn’t be allowed, but I’d like your thoughts @blakerouse @ycombinator

I don't think there is really away of having Fleet prevent this. I could see two requests coming in near the same time one scheduled and the other not. To me the non-scheduled wins and should just happen, the scheduled one should just be ignored as the upgrade is now happening.
